src/btree/row_key.c

```diff
@@ -420,6 +420,14 @@ __wt_row_ikey(
          */
         WT_ASSERT(session, oldv == 0 || (oldv & WT_IK_FLAG) != 0);
         WT_ASSERT(session, ref->state != WT_REF_SPLIT);
+        /*
+         * FIXME-WT-10044 - This is an example of coupling between our previous implementation of
+         * WT_ASSERT and HAVE_DIAGNOSTIC. We previously assumed that WT_ASSERT code always ran when
+         * HAVE_DIAGNOSTIC=1, but now it's possible to turn off diagnostic_asserts via a session or
+         * connection reconfigure, which in this case causes a segfault due to the CAS below not
+         * running. This is a rare case though - we're highly unlikely to ever turn off diagnostic
+         * asserts while also compiling with HAVE_DIAGNOSTIC=1.
```

Comment:
This makes me a little uneasy - you're right about this being rare, but I'm wondering if we can eliminate it until WT-10044 is sorted.

As things stand right now, is there any legit use for building with `HAVE_DIAGNOSTIC=1` then configuring it off? If not, perhaps asserting that it's never the case and crashing from the config function would be a good option, so nobody is trying to track down mystery segfaults in case their config gets messed up.

Also, the ticket mentions spotting these issues by changing `WT_ASSERT` to run in release mode, does that spot the case above where there's code that just doesn't run?

Reply:
This is on a feature branch, so I think we'll be ok fixing this in WT-10044 before its hits mainline.
Whether we even want this behaviour is a good question.. I could see it making more sense to force asserts on in diagnostic mode. 

This wouldn't be spotted by running WT_ASSERT in release mode, but I can confirm this is the only case as I checked all the other asserts when finding this issue. I think our core issue is running code with side affects inside assertions but I'll brainstorm on what we do about this in WT_10044.

Reply:
Yep, fair enough. Thanks for clarifying that.